creta_bata.py

```diff
@@ -0,0 +1,427 @@
+    
+    errPrev = 1
+    errCurr = 0
+    ii = 1
+    
+    while ii <= param['numofouteriter'] and abs(errCurr - errPrev) / errPrev > 1e-5:
+        Aij = At0 @ t
+        tij_T_weighted = tij_observe.flatten()
+        Svec = np.sum(Aij ** 2, axis=0) / np.sum(Aij * tij_T_weighted.reshape(3, -1), axis=0)
+        tmp3 = np.tile(Svec, 3)
+        S = tmp3
+        
+        A = csr_matrix(np.diag(1 / S)) * At0
+        B = tij_observe.flatten()
+        tmp = np.sqrt(np.sum((A @ t - B) ** 2, axis=0))
+        errPrev = errCurr
+        errCurr = np.sum(tmp[ed_ret_idx.flatten()])
+        Wvec = 1 / (1 + ((tmp / param['robustthre']) ** 2))
+        
+        jj = 1
+        while jj <= param['numofinneriter']:
+            Aij = At0 @ t
+            tij_T_weighted = tij_observe.flatten()
+            Svec = np.sum(Aij ** 2, axis=0) / np.sum(Aij * tij_T_weighted.reshape(3, -1), axis=0)
+            
+            ed_ret_idx, node_ret_idx_temp = extractLargestConnComp(Svec, tij_index, node_ret_idx)
+            node_ret_idx_temp2 = node_ret_idx_temp[:, np.setdiff1d(np.arange(node_ret_idx_temp.shape[1]), idxConst - 1)]
+            Svec = Svec[ed_ret_idx.flatten()]
+            tmp3 = np.tile(Svec, 3)
+            S_red = tmp3
+            Wvec2 = Wvec[ed_ret_idx.flatten()]
+            At0_fullr = At0_full[ed_ret_idx.flatten(), :][:, node_ret_idx_temp2.flatten()]
+            tij_T = tij_observe[:, ed_ret_idx.flatten()]
+            W = np.tile(np.sqrt(Wvec2), 3)
+            
+            A = csr_matrix(np.diag(W / S_red)) * At0_fullr
+            B = W * tij_T.flatten()
+            t2 = spsolve(A.T @ A, A.T @ B)
+            t[node_ret_idx_temp2.flatten()] = t2
+            jj += 1
+        
+        ii += 1
+    
+    t = t.reshape(3, -1)
+    t[:, :idxConst - 1] = t[:, np.arange(idxConst - 1) - 1]
+    t[:, idxConst - 1:] = t[:, np.arange(idxConst - 1, t.shape[1] - 1)]
+    t[~node_ret_idx.flatten()] = np.nan
+    t = t - np.nanmean(t, axis=1, keepdims=True)
+    nodes = np.where(node_ret_idx[0, :])[0]
+    ed_ret_idx = np.isin(tij_index[0, :], nodes) & np.isin(tij_index[1, :], nodes)
+    
+    return t, ed_ret_idx
+
+
+def extractLargestConnComp(Svec, tij_index, node_ret_idx):
+    ed_ret_idx = np.ones((3, tij_index.shape[1]), dtype=bool)
+    ed_ret_idx[:, Svec < 0] = False
+    nodes = np.where(node_ret_idx[0])[0]
+    eidx = np.isin(tij_index[0], nodes) & np.isin(tij_index[1], nodes)
+    ed_ret_idx[:, ~eidx] = False
+    edges = tij_index[:, ed_ret_idx[0]]
+
+    G = nx.Graph()
+    G.add_edges_from(edges.T)
+    bins = list(nx.connected_components(G))
+    largest_comp = np.array(list(max(bins, key=len))).astype(int)
+    ed_ret_idx_temp = np.isin(tij_index[0], largest_comp) & np.isin(tij_index[1], largest_comp) & ed_ret_idx[0]
+    ed_ret_idx = np.tile(ed_ret_idx_temp, (3, 1))
+    node_ret_idx[...] = False
+    node_ret_idx[:, largest_comp] = True
+
+    return ed_ret_idx, node_ret_idx
+
+
+
+extractParallelRigidSubgraph = True
```

Comment:
could we put L270 to L424 in a function `def run_creta()` or `def creta_demo()`?